[Layout] Layout rebuilds after resize/reposition #32

akhenry opened the pull request:
Summary of changes:
* Layout view no longer `uses` composition capability. This was triggering unnecessary redraws of all composed objects on any model change. See https://github.com/nasa/openmctweb/issues/32#issuecomment-158566829
* `LayoutController` now fetches composition from compositionCapability on domain object, but only when model.composition changes.

### Author checklist
* Changes address original issue? __Y__
* Unit tests included and/or updated with changes? __Y__
* Command line build passes? __Y__
* Expect to pass code review? __Y__

VWoeltjen commented on platform/features/layout/src/LayoutController.js:
```diff
@@ -58,14 +58,14 @@ define(
+             * Defined as member function to facilitate testing.
+             * @private
+             */
+            LayoutController.prototype.layoutPanels = function layoutPanels (ids) {
```
This assignment is rather funny since it's in the constructor for LayoutController; this method gets reassigned with a new closure scope every time a new LayoutController gets instantiated.

I would do one of:

1. Move outside of the constructor call; may need to store some additional properties on `this` at constructor time to support this
2. Assign as `this.layoutPanels = ` such that this varies per instance (and can still access `$scope` et al via closure scope.)
3. Just leave it as an inner function (as `lookupPanels` had been)

akhenry replied:
Ah, whoops, this was by accident. Thanks for picking it up. I promoted an inner function to a member function so that I could use it in testing, but modified it in place by mistake. With long constructors it can be difficult to know whether you're currently inside or outside of the constructor.

VWoeltjen commented on platform/features/layout/src/LayoutController.js:
```diff
@@ -156,8 +175,8 @@ define(
-            // Position panes when the model field changes
-            $scope.$watch("model.composition", lookupPanels);
+            // Update composed objects on screen, and position panes
+            $scope.$watch("model.composition", refreshComposition);
```
I believe a `$watch` here would fail to detect changes made in-place to `model.composition`. `$watchCollection` would be more appropriate here, although for large compositions it could be costly.

If `$watchCollection` is too heavy, you could use a mutation listener and explicitly compare the `composition` array to see if there are any changes.

akhenry replied:
Good catch! Changed it to $watchCollection. Oddly, this was working with $watch (I tested a composition change to ensure it actually gets triggered). Possibly what's happening is that somewhere along the line the .composition attribute of the model gets replaced anyway on update. But $watchCollection is definitely more correct.

VWoeltjen commented on platform/features/layout/src/LayoutController.js:
```diff
@@ -127,6 +127,25 @@ define(
+                return getComposition($scope.domainObject)
+                    .then(composeView)
+                    .then(self.layoutPanels);
+            }
```
`refreshComposition` could have `ids` available to it as an argument from the watch - could simplify code a little bit and remove some of the above functions if we used these to `layoutPanels` instead of mapping them back from domain objects in `composeView`.

There's also a race condition introduced here, since `composition` promises will not necessarily resolve in the same order they are requested. For instance, consider the following sequence:

* A layout is loaded containing composition A, B, C. A request is issued but it doesn't resolve right away (for whatever reason)
* From the tree, the user creates a Link, adding D to the composition of the layout.
* A new `refreshComposition` call loads A, B, C, D. This happens to resolve quickly; A, B, C, D are placed in scope.
* The old composition promise finally resolves; A, B, C is put into scope.
* Digest gets triggered, user sees A, B, C, thinks "where is that D that I just added?"

There's [some logic in `mct-representation` that prevents this](https://github.com/nasa/openmctweb/blob/560454e7c253b96c0bf461d0c2c7e336be17f560/platform/representation/src/MCTRepresentation.js#L127-L132) for capabilities loaded via `uses`, so this would be a regression.

akhenry replied:
refreshComposition could have ids available to it as an argument from the watch

I know, it's a bit icky. The reasoning behind this is actually tangentially related to the issue below. Basically I ignore the ids from the watch, because I subsequently fetch the composition from a capability (because it's no longer provided on `$scope`), and there is no guarantee the ids will match. So instead I just use the ids returned from the capability since that is the composition that is actually displayed in the layout. 

    There's some logic in mct-representation that prevents this for capabilities loaded via uses, so    this would be a regression.

Aha! I saw that bit of code and wondered what was going on there :) The comment said something about Angular reusing scope and subsequent representations, but I didn't realize it was relevant to this situation. 

It looks like the logic is relatively straightforward, but will need to be implemented carefully. I'll try to write a test scenario that verifies the fix as well.

VWoeltjen replied:
> Basically I ignore the ids from the watch, because I subsequently fetch the composition from a capability (because it's no longer provided on $scope), and there is no guarantee the ids will match. So instead I just use the ids returned from the capability since that is the composition that is actually displayed in the layout.

Ah, that makes sense. While I _think_ it would be safe to assume IDs won't change (or that the watch will be invoked again when they do), agree that the approach you're using makes it easier to be confident of correctness.

VWoeltjen commented:
Added some line comments. I think this is the right approach - it makes me wonder if we should continue to maintain the `uses` property of `representations` at all. Glad that we will have a fix for this issue; it's mostly cosmetic but I find it rather irritating.

larkin commented:
I like the idea to remove 'used' from representation. Moves logic to controller and it's clearer to see where scope variables came from. More so, it promotes destroying scopes instead of reusing them, so that we can use standard scope lifecycle handling.

akhenry commented:
Addressed all of the comments. Specific comments inline.

Have added a counter ala MctRepresentation that ensures that stale callbacks are discarded

VWoeltjen commented:
Changes address comments, proceeding with merge.

### Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y
